tsl/src/compression/compression.c

```diff
@@ -2042,3 +2047,226 @@ update_compressed_chunk_relstats(Oid uncompressed_relid, Oid compressed_relid)
+			 * constraints.
+			 *
+			 * 1. Column is segmentby-Column
+			 * In this case we can add a single ScanKey with an
+			 * equality check for the value.
+			 * 2. Column is orderby-Column
+			 * In this we can add 2 ScanKeys with range constraints
+			 * utilizing batch metadata.
+			 * 3. Column is neither segmentby nor orderby
+			 * In this case we cannot utilize this column for
+			 * batch filtering as the values are compressed and
+			 * we have no metadata.
+			 */
+
+			if (COMPRESSIONCOL_IS_SEGMENT_BY(fd))
+			{
+				bool isnull;
+				Datum value = slot_getattr(slot, attno, &isnull);
+
+				TypeCacheEntry *tce =
+					lookup_type_cache(decompressor.in_desc->attrs[attno - 1].atttypid,
+									  TYPECACHE_EQ_OPR_FINFO);
+
+				if (!tce || !OidIsValid(tce->eq_opr_finfo.fn_oid))
+					elog(ERROR,
+						 "no equality function for type \"%s\"",
+						 format_type_be(decompressor.in_desc->attrs[attno - 1].atttypid));
+
+				/*
+				 * In PG versions <= 14 NULL values are always considered distinct
+				 * from other NULL values and therefore NULLABLE multi-columnn
+				 * unique constraints might expose unexpected behaviour in the
+				 * presence of NULL values.
+				 * Since SK_SEARCHNULL is not supported by heap scans we cannot
+				 * build a ScanKey for NOT NULL and instead have to do those
+				 * checks manually.
+				 */
+				if (isnull)
+				{
+					*null_columns = bms_add_member(*null_columns, attno);
+				}
+				else
+				{
+					ScanKeyEntryInitialize(&scankeys[index],
+										   0, /* flags */
+										   attno,
+										   BTEqualStrategyNumber,
+										   InvalidOid, /* No strategy subtype. */
+										   decompressor.in_desc->attrs[attno - 1].attcollation,
+										   tce->eq_opr_finfo.fn_oid,
+										   value);
+					index++;
+				}
+			}
+		}
+	}
+
+	*num_scankeys = index;
+	return scankeys;
+}
+
+void
+decompress_batches_for_insert(Chunk *chunk, TupleTableSlot *slot)
+{
+	Relation out_rel = table_open(chunk->table_id, RowExclusiveLock);
+
+	if (!ts_indexing_relation_has_primary_or_unique_index(out_rel))
+	{
+		/*
+		 * If there are no unique constraints there is nothing to do here.
+		 */
+		table_close(out_rel, RowExclusiveLock);
+		return;
+	}
+
+	Chunk *comp = ts_chunk_get_by_id(chunk->fd.compressed_chunk_id, true);
+	Relation in_rel = relation_open(comp->table_id, RowExclusiveLock);
+
+	RowDecompressor decompressor = build_decompressor(in_rel, out_rel);
+	Bitmapset *unique_columns = relation_unique_columns(out_rel);
+	Bitmapset *null_columns = NULL;
+
+	int num_scankeys;
+	ScanKeyData *scankeys = build_scankeys(chunk->fd.hypertable_id,
+										   decompressor,
+										   unique_columns,
+										   &null_columns,
+										   slot,
+										   &num_scankeys);
+
+	TableScanDesc heapScan =
+		table_beginscan(in_rel, GetTransactionSnapshot(), num_scankeys, scankeys);
+
+	for (HeapTuple compressed_tuple = heap_getnext(heapScan, ForwardScanDirection);
```

Comment:
Heap scans like this without a per-tuple memory context make me suspicious. I worry that there's a risk this allocates memory proportional to the size of the table. You should double check the memory usage, especially for big tables.

Reply:
the row decompressor switches to a per tuple memory context